### PR TITLE
#5304 Use listeners copy to prevents concurrent modifications of the listeners list

### LINKED ISSUE #5304: ConcurrentModificationException in ReplayLocationEngine
While updating listeners with replayed locations there is possibility that a listener decides to stop receiving updates and removes itself from `registeredCallbacks`. In that case ConcurrentModificationException will happen.

https://github.com/mapbox/mapbox-navigation-android/blob/main/libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/ReplayLocationEngine.kt#L109

**Mapbox Navigation SDK version:** 2.2.0-alpha.1

### PR DESCRIPTION
**Pull request by evabishchevich**
Fixes #5304 
### Description

This PR fixes ConcurrentModificationException on the location callbacks list.

### CODE REVIEW

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/ReplayLocationEngine.kt [hunk 1]
```diff
@@ -106,7 +106,7 @@ class ReplayLocationEngine(
         val locationEngineResult = LocationEngineResult.create(location)
         lastLocationEngineResult = locationEngineResult
 
-        registeredCallbacks.forEach { it.onSuccess(locationEngineResult) }
+        registeredCallbacks.toTypedArray().forEach { it.onSuccess(locationEngineResult) }
```
Comment (LukasPaczos) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/ReplayLocationEngine.kt:109]:
```suggestion
        registeredCallbacks.toTypedArray().forEach { it.onSuccess(locationEngineResult) }
```
We generally use `CopyOnWriteArrayList`s to support this use case across the SDK, without having to remember to copy the list on each invocation. Could you refactor the field itself?

Comment (evabishchevich) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/ReplayLocationEngine.kt:109]: sure thing. It was a second option

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/ReplayLocationEngine.kt [hunk 2]
```diff
@@ -22,7 +23,7 @@ class ReplayLocationEngine(
     mapboxReplayer: MapboxReplayer
 ) : LocationEngine, ReplayEventsObserver {
 
-    private val registeredCallbacks: MutableList<EngineCallback> = mutableListOf()
+    private val registeredCallbacks: MutableList<EngineCallback> = CopyOnWriteArrayList()
     private val lastLocationCallbacks: MutableList<EngineCallback> = mutableListOf()
```
Comment (Zayankovsky) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/ReplayLocationEngine.kt:27]: I think we should do the same for `lastLocationCallbacks`.

**Review by Zayankovsky — APPROVED** [review]

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/ReplayLocationEngine.kt (hunk 2 above)
Comment (evabishchevich) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/ReplayLocationEngine.kt:27]: no need as there is no API to remove an item from `lastLocationCallbacks` from the outside

**Comment by LukasPaczos** [comment]
Last thing - we're just missing a changelog entry (CI job failing). Could you draft one in `CHANGELOG.md`?

File: CHANGELOG.md [hunk 3]
```diff
@@ -3,6 +3,7 @@
 Mapbox welcomes participation and contributions from everyone.
 
 ## Unreleased
+- Fixed: crash when location listener removes itself during getting a location update. [#5305](https://github.com/mapbox/mapbox-navigation-android/pull/5305)
```
Comment (LukasPaczos) [inline CHANGELOG.md:6]:
```suggestion
- Fixed a crash when `ReplayLocationEngine`'s location callback removes itself during getting a location update. [#5305](https://github.com/mapbox/mapbox-navigation-android/pull/5305)
```

You'll need to push to resolve the changelog entry conflict anyway, sorry! We're working on improving that.

**Review by LukasPaczos — APPROVED** [review]

**Comment by LukasPaczos** [comment]
You'll need to rebase on `main` to get the test to pass :upside_down_face:

**Pull request merged**